File: app/src/main/java/com/androlua/LuaApplication.java
package com.androlua;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Environment;
import androidx.preference.PreferenceManager;
import com.google.android.material.color.DynamicColors;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import org.luaj.Globals;
import org.luaj.LuaTable;

/** Created by nirenr on 2019/12/13. */
public class LuaApplication extends Application implements LuaContext {
  private static LuaApplication instance;
  private String mExtDir;
  private static final HashMap sGlobalData = new HashMap();

  public static LuaApplication getInstance() {
    return instance;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    instance = this;
    rmDir(getExternalFilesDir("dexfiles"));
    CrashHandler.getInstance().init(this);
    DynamicColors.applyToActivitiesIfAvailable(this);
  }

  public static boolean rmDir(File dir) {
    if (dir.isDirectory()) {
      File[] fs = dir.listFiles();
      for (File f : fs) rmDir(f);
    }
    return dir.delete();
  }

  @Override
  public ArrayList<ClassLoader> getClassLoaders() {
    return null;
  }

  @Override
  public void call(String func, Object... args) {}

  @Override
  public void set(String name, Object value) {}

  @Override
  public String getLuaPath() {
    return null;
  }

  public String getLuaPath(String s) {
    return new File(getLuaDir(), s).getAbsolutePath();
  }

  @Override
  public String getLuaPath(String dir, String name) {
    return null;
  }

  public String getLuaDir() {
    return getFilesDir().getAbsolutePath();
  }

  @Override
  public String getLuaDir(String dir) {
    return null;
  }

  public String getLuaExtDir() {
    if (mExtDir != null) return mExtDir;
    File d = new File(Environment.getExternalStorageDirectory(), "LuaJ");
    if (!d.exists()) d.mkdirs();
    mExtDir = d.getAbsolutePath();
    return mExtDir;
  }

  @Override
  public String getLuaExtDir(String dir) {
    File d = new File(getLuaExtDir(), dir);
    if (!d.exists()) d.mkdirs();
    return d.getAbsolutePath();
  }

  @Override
  public void setLuaExtDir(String dir) {
    mExtDir = dir;
  }

  @Override
  public String getLuaExtPath(String path) {
    return new File(getLuaExtDir(), path).getAbsolutePath();
  }

  @Override
  public String getLuaExtPath(String dir, String name) {
    return new File(getLuaExtDir(dir), name).getAbsolutePath();
  }

  @Override
  public Context getContext() {
    return this;
  }

  @Override
  public Globals getLuaState() {
    return null;
  }

  @Override
  public Object doFile(String path, Object... arg) {
    return null;
  }

  @Override
  public void sendMsg(String msg) {}

  @Override
  public void sendError(String title, Exception msg) {}

  @Override
  public int getWidth() {
    return 0;
  }

  @Override
  public int getHeight() {
    return 0;
  }

  @Override
  public Map getGlobalData() {
    return sGlobalData;
  }

  @Override
  public Map<String, ?> getSharedData() {
    return PreferenceManager.getDefaultSharedPreferences(this).getAll();
  }

  @Override
  public Object getSharedData(String key) {
    return PreferenceManager.getDefaultSharedPreferences(this).getAll().get(key);
  }

  @Override
  public Object getSharedData(String key, Object def) {
    Object ret = PreferenceManager.getDefaultSharedPreferences(this).getAll().get(key);
    if (ret != null) return ret;
    return def;
  }

  @Override
  public boolean setSharedData(String key, Object value) {
    SharedPreferences.Editor edit = PreferenceManager.getDefaultSharedPreferences(this).edit();
    if (value == null) edit.remove(key);
    else if (value instanceof String) edit.putString(key, value.toString());
    else if (value instanceof Long) edit.putLong(key, (Long) value);
    else if (value instanceof Integer) edit.putInt(key, (Integer) value);
    else if (value instanceof Float) edit.putFloat(key, (Float) value);
    else if (value instanceof LuaTable)
      edit.putStringSet(key, new HashSet(((LuaTable) value).values()));
    else if (value instanceof Set) edit.putStringSet(key, (Set<String>) value);
    else if (value instanceof Boolean) edit.putBoolean(key, (Boolean) value);
    else return false;
    return edit.commit();
  }

  @Override
  public void regGc(LuaGcable obj) {}

  @Override
  public InputStream findResource(String name) {
    try {
      if (new File(name).exists()) return new FileInputStream(name);
    } catch (Exception e) {
      /*
      e.printStackTrace();*/
    }
    try {
      return new FileInputStream(new File(getLuaPath(name)));
    } catch (Exception e) {
      /*
      e.printStackTrace();*/
    }
    try {
      return getAssets().open(name);
    } catch (Exception ioe) {
      /*
      e.printStackTrace();*/
    }
    return null;
  }

  public boolean checkResource(String name) {
    try {
      if (new File(name).exists()) return true;
    } catch (Exception e) {

    }
    try {
      return new File(getLuaPath(name)).exists();
    } catch (Exception e) {
      /*
      e.printStackTrace();*/
    }
    try {
      InputStream in = getAssets().open(name);
      in.close();
      return true;
    } catch (Exception ioe) {

    }
    return false;
  }

  @Override
  public String findFile(String filename) {
    if (filename.startsWith("/")) return filename;
    return getLuaPath(filename);
  }
}

File: app/src/main/java/com/androlua/LuaApplication.kt
package com.androlua

import android.app.Application
import android.content.Context
import android.os.Environment
import androidx.preference.PreferenceManager
import com.nekolaska.internal.commit
import org.luaj.Globals
import org.luaj.LuaTable
import java.io.File
import java.io.FileInputStream

/**
 * Created by nirenr on 2019/12/13.
 */
class LuaApplication : Application(), LuaContext {
    private var mExtDir: String? = null
    override fun onCreate() {
        super.onCreate()
        instance = this
        getExternalFilesDir("dexfiles")?.let { rmDir(it) }
        CrashHandler.instance.init(this)
        //DynamicColors.applyToActivitiesIfAvailable(this)
        //EmulatorSDK.init(this)
    }

    override fun getClassLoaders(): ArrayList<ClassLoader>? {
        return null
    }

    override fun call(func: String, vararg args: Any) {
    }

    override fun set(name: String, value: Any) {
    }

    override fun getLuaPath(): String {
        return filesDir.absolutePath
    }

    override fun getLuaPath(s: String): String {
        return filesDir.resolve(s).absolutePath
    }

    override fun getLuaPath(dir: String, name: String): String? {
        return filesDir.resolve(dir).resolve(name).absolutePath
    }

    override fun getLuaDir(): String {
        return filesDir.absolutePath
    }

    override fun getLuaDir(dir: String): String? {
        return filesDir.resolve(dir).absolutePath
    }

    override fun getLuaExtDir(): String {
        if (mExtDir != null) return mExtDir as String
        val d = File(Environment.getExternalStorageDirectory(), "LuaJ")
        if (!d.exists()) d.mkdirs()
        mExtDir = d.absolutePath
        return mExtDir as String
    }

    override fun getLuaExtDir(dir: String): String {
        val d = File(luaExtDir, dir)
        if (!d.exists()) d.mkdirs()
        return d.absolutePath
    }

    override fun setLuaExtDir(dir: String) {
        mExtDir = dir
    }

    override fun getLuaExtPath(path: String): String {
        return File(luaExtDir, path).absolutePath
    }

    override fun getLuaExtPath(dir: String, name: String): String {
        return File(getLuaExtDir(dir), name).absolutePath
    }

    override fun getContext(): Context {
        return this
    }

    override fun getLuaState(): Globals? {
        return null
    }

    override fun doFile(path: String, vararg arg: Any): Any? {
        return null
    }

    override fun sendMsg(msg: String) {
    }

    override fun sendError(title: String, msg: Exception) {
    }

    override fun getWidth(): Int {
        return 0
    }

    override fun getHeight(): Int {
        return 0
    }

    override fun getGlobalData(): Map<*, *> {
        return sGlobalData
    }

    override fun getSharedData(): Map<String, *> {
        return PreferenceManager.getDefaultSharedPreferences(this).all
    }

    override fun getSharedData(key: String): Any {
        return PreferenceManager.getDefaultSharedPreferences(this).all[key]!!
    }

    override fun getSharedData(key: String, def: Any): Any {
        val ret = PreferenceManager.getDefaultSharedPreferences(this).all[key]
        if (ret != null) return ret
        return def
    }

    @Suppress("UNCHECKED_CAST")
    override fun setSharedData(key: String, value: Any?): Boolean {
        return PreferenceManager.getDefaultSharedPreferences(this).commit {
            if (value == null)
                remove(key)
            else when (value.javaClass.simpleName) {
                "String" -> putString(key, value as String)
                "Long" -> putLong(key, (value as Long))
                "Integer" -> putInt(key, (value as Int))
                "Float" -> putFloat(key, (value as Float))
                "LuaTable" -> putString(
                    key,
                    (value as LuaTable).values().toString()
                )

                "Set" -> putStringSet(key, value as Set<String>)
                "Boolean" -> putBoolean(key, (value as Boolean))
                else -> return false
            }
        }
    }


    override fun regGc(obj: LuaGcable) {
    }

    override fun findResource(name: String): FileInputStream? {
        try {
            if (File(name).exists()) return FileInputStream(name)
        } catch (_: Exception) {
            /*
      e.printStackTrace();*/
        }
        try {
            return FileInputStream(getLuaPath(name))
        } catch (_: Exception) {
            /*
      e.printStackTrace();*
        }
        try {
            return getAssets().open(name);
        } catch (Exception ioe) {
      / *
      e.printStackTrace();*/
        }
        return null
    }

    fun checkResource(name: String): Boolean {
        try {
            if (File(name).exists()) return true
        } catch (_: Exception) {
        }
        try {
            return File(getLuaPath(name)).exists()
        } catch (_: Exception) {
        }
        try {
            val `in` = assets.open(name)
            `in`.close()
            return true
        } catch (_: Exception) {
        }
        return false
    }

    override fun findFile(filename: String): String {
        if (filename.startsWith("/")) return filename
        return getLuaPath(filename)
    }


    companion object {
        @JvmStatic
        lateinit var instance: LuaApplication

        private val sGlobalData: HashMap<*, *> = HashMap<Any?, Any?>()

        fun rmDir(dir: File): Boolean {
            if (dir.isDirectory) {
                for (f in dir.listFiles()!!) rmDir(f)
            }
            return dir.delete()
        }
    }
}
